pynbody/array/__init__.py

```diff
@@ -400,15 +400,27 @@ def __array_function__(self, func, types, args, kwargs):
         result = func(*args_processed, **kwargs_processed)
 
         if func in SimArray._ufunc_registry:
-            result = result.view(SimArray)
-            if isinstance(result, SimArray): # may not be true if the result is a scalar
-                sim = None
-                for arg in args:
-                    if isinstance(arg, SimArray):
-                        sim = arg.sim
-                        break
-                result.units = SimArray._ufunc_registry[func](*args, **kwargs)
-                result.sim = sim
+            
+            sim = None
+            for arg in args:
+                if isinstance(arg, SimArray):
+                    sim = arg.sim
+                    break
+            units_val = SimArray._ufunc_registry[func](*args, **kwargs)
+            
+            # If result is a tuple, convert each element
+            if isinstance(result, tuple):
+                result = tuple( r.view(SimArray) for r in result)
+                result = tuple(
```

Comment:
While this comprehension is quite clever, I think it's also a bit too cryptic. Also, it means duplicating essentially the same logic used for a single array below. I'd rather have something like

```python
def _as_sim_array(result, units, sim):
   result = result.view(SimArray)
   result.units = units
   result.sim = sim

...


if isinstance(result, tuple):
   result = tuple(_as_sim_array(r_i, units_val, sim) for r_i in result)
else:
   result = _as_sim_array(result, units_val, sim)
```